### ISSUE TITLE
Async code in custom rules?

### ISSUE THREAD

**Issue by katyaaronov** [vsiakka/gherkin-lint#342]
Hello,

Should it be possible to use asynchronous code in custom rules?

For example, I want to enforce every scenario to have a tag which corresponds to a Jira issue of type Story. 

For this I iterate through all scenarios and their tags and run axios.get() to send API request to get an issue type for every tag. 

axios.get() returns a promise and I am unable to get it resolved - the .then/.catch code is not even being reached, so I wonder if this is even expected to be supported.

P.S. tried it with a simple setTimeout promise with no luck as well.

Thanks,
Katya.

**Comment by katyaaronov** [comment]
Digging deeper into source code I found out that resolving promises returned by run() functions are not being awaited.
I made the following changes locally to check if it resolves the issue and it seems that it does.
Please consider updating the source code with the below adjustments.

1. _**src/main.js:**_
Make the linter.lint() callback function async:
```
linter.lint(files, config, additionalRulesDirs).then(async results => {
```

2. **_src/linter.js:_**
Add async/await in the following code in lint():
```
    async ({
             feature,
             file
           }) => {
       perFileErrors = await rules.runAllEnabledRules(feature, file, configuration, additionalRulesDirs);
```

3. **_src/rules.js:_**
Replace `forEach` (which does not wait for promises to resolve) by `for...of` (which handle asynchronous tasks) and add async/await in runAllEnabledRules():
```
async function runAllEnabledRules(feature, file, configuration, additionalRulesDirs) {
  let errors = [];
  const rules = getAllRules(additionalRulesDirs);
  for (const ruleName of Object.keys(rules)) {
    let rule = rules[ruleName];

    if (isRuleEnabled(configuration[rule.name])) {
      const ruleConfig = Array.isArray(configuration[rule.name]) ? configuration[rule.name][1] : {};
      const error = await rule.run(feature, file, ruleConfig);
```

Regards,
Katya.

**Comment by karoisuzvards** [comment]
I stumbled across this issue too.

Instead of patching **gherkin-linter** I used:

* **fetch API** (added to **Node.js** in version **18**, released in **April 2022**)
* [sync-fetch package](https://github.com/larsgw/sync-fetch#readme)

```js
const fetch = require('sync-fetch')

const metadata = fetch('https://jira/issue/v2/yadayada', {
  headers: {
    Accept: 'application/vnd.citationstyles.csl+json',
    ...
  }
}).json()
```
